server/driver/wivrn_foveation.cpp

```diff
@@ -0,0 +1,279 @@
+		                                                          },
+		                                                          .layout = *layout,
+		                                                  });
+	}
+
+	std::array<vk::DescriptorSetLayout, 2> layouts = {*ds_layout, *ds_layout};
+
+	auto my_ds = vk.device.allocateDescriptorSets({
+	        .descriptorPool = *dp,
+	        .descriptorSetCount = 2,
+	        .pSetLayouts = layouts.data(),
+	});
+
+	ds = {
+	        my_ds[0].release(),
+	        my_ds[1].release(),
+	};
+}
+
+void wivrn_foveation_renderer::render_distortion_images(std::array<to_headset::foveation_parameter, 2> foveation_arr, const VkImage * images, const VkImageView * image_views)
+{
+	cmd_buf.reset();
+	cmd_buf.begin(vk::CommandBufferBeginInfo{});
+
+	// Monado's distortion shader is patched to only consider the red channel's image
+	for (int eye = 0; eye < 2; ++eye)
+	{
+		// Descriptor set
+
+		vk::DescriptorImageInfo image_info{
+		        .imageView = image_views[eye],
+		        .imageLayout = vk::ImageLayout::eGeneral,
+		};
+		vk.device.updateDescriptorSets(
+		        vk::WriteDescriptorSet{
+		                .dstSet = ds[eye],
+		                .dstBinding = 0,
+		                .descriptorCount = 1,
+		                .descriptorType = vk::DescriptorType::eStorageImage,
+		                .pImageInfo = &image_info,
+		        },
+		        nullptr);
+
+		std::array im_barriers0 = {
+		        vk::ImageMemoryBarrier{
+		                .srcAccessMask = vk::AccessFlagBits::eNone,
+		                .dstAccessMask = vk::AccessFlagBits::eMemoryWrite,
+		                .oldLayout = vk::ImageLayout::eShaderReadOnlyOptimal,
+		                .newLayout = vk::ImageLayout::eGeneral,
+		                .image = images[eye],
+		                .subresourceRange = {.aspectMask = vk::ImageAspectFlagBits::eColor,
+		                                     .baseMipLevel = 0,
+		                                     .levelCount = 1,
+		                                     .baseArrayLayer = 0,
+		                                     .layerCount = 1},
+		        },
+		};
+		cmd_buf.pipelineBarrier(
+		        vk::PipelineStageFlagBits::eTopOfPipe,
+		        vk::PipelineStageFlagBits::eComputeShader,
+		        {},
+		        nullptr,
+		        nullptr,
+		        im_barriers0);
+
+		auto foveation = foveation_arr[eye];
+
+		FoveationParamsPcs f_params{
+		        .a = {foveation.x.a, foveation.y.a},
+		        .b = {foveation.x.b, foveation.y.b},
+		        .scale = {foveation.x.scale, foveation.y.scale},
+		        .center = {foveation.x.center, foveation.y.center},
+		};
+
+		cmd_buf.bindPipeline(vk::PipelineBindPoint::eCompute, *pipeline);
+		cmd_buf.bindDescriptorSets(vk::PipelineBindPoint::eCompute, *layout, 0, ds[eye], {});
+		cmd_buf.pushConstants<FoveationParamsPcs>(*layout, vk::ShaderStageFlagBits::eCompute, 0, f_params);
+		cmd_buf.dispatch(16, 16, 1);
```

Comment:
The 16s here are 128 (from image size constant in monado's code) / 8 (from shader local size). Can we use the constant from monado?
For the 8 from shader, as long as it's under WiVRn's control I don't care if it's a magic value.

Reply:
i couldn't include `render/render_interface.h` because then it complains about:

`In included file: unknown type name 'PFN_vkCreateXcbSurfaceKHR'; did you mean 'PFN_vkDestroySurfaceKHR'?`

Reply:
in the shader I can do 
`float one_over_dim = 1.0 / float((gl_WorkGroupSize.x * gl_NumWorkGroups.x) - 1);`
but then it won't be a compile-time const